packages/evm/src/eof/verify.ts

```diff
@@ -0,0 +1,519 @@
+  const containerTypeMap = new Map<number, ContainerSectionType>()
+
+  function addJump(location: number) {
+    if (intermediateBytes.has(location)) {
+      // When trying to JUMP into an intermediate byte: this is invalid
+      validationError(EOFError.InvalidRJUMP)
+    }
+    jumpLocations.add(location)
+  }
+
+  function addIntermediate(location: number) {
+    if (jumpLocations.has(location)) {
+      // When trying to add an intermediate to a location already JUMPed to: this is invalid
+      validationError(EOFError.InvalidRJUMP)
+    }
+    intermediateBytes.add(location)
+  }
+
+  // TODO (?) -> stackDelta currently only has active EOF opcodes, can use it directly (?)
+  // (so no need to generate the valid opcodeNumbers)
+
+  // Validate each code section
+  const opcodes = evm.getActiveOpcodes()
+
+  const opcodeNumbers = new Set<number>()
+
+  for (const [key] of opcodes) {
+    opcodeNumbers.add(key)
+  }
+
+  // Add INVALID as valid
+  opcodeNumbers.add(0xfe)
+
+  // Remove CODESIZE, CODECOPY, EXTCODESIZE, EXTCODECOPY, EXTCODEHASH, GAS
+  opcodeNumbers.delete(0x38)
+  opcodeNumbers.delete(0x39)
+  opcodeNumbers.delete(0x5a)
+  opcodeNumbers.delete(0x3b)
+  opcodeNumbers.delete(0x3c)
+  opcodeNumbers.delete(0x3f)
+
+  // Remove CALLCODE and SELFDESTRUCT
+  opcodeNumbers.delete(0xf2)
+  opcodeNumbers.delete(0xff)
+
+  // TODO omnibus https://github.com/ipsilon/eof/blob/main/spec/eof.md states
+  // JUMP / JUMPI / PC / CREATE / CREATE2 also banned
+  // This is not in the EIPs yet
+  // Add these opcodes here
+
+  opcodeNumbers.delete(0x56) // JUMP
+  opcodeNumbers.delete(0x57) // JUMPI
+
+  opcodeNumbers.delete(0x58) // PC
+
+  opcodeNumbers.delete(0xf0) // CREATE
+  opcodeNumbers.delete(0xf5) // CREATE2
+
+  // Note: this name might be misleading since this is the list of opcodes which are OK as final opcodes in a code section
+  // TODO if using stackDelta for EOF it is possible to add a "termination" boolean for the opcode to mark it as terminating
+  // (so no need to generate this set here)
+  const terminatingOpcodes = new Set<number>()
+
+  terminatingOpcodes.add(0x00) // STOP
+  terminatingOpcodes.add(0xf3) // RETURN
+  terminatingOpcodes.add(0xfd) // REVERT
+  terminatingOpcodes.add(0xfe) // INVALID
+
+  terminatingOpcodes.add(0xee) // RETURNCONTRACT
+
+  terminatingOpcodes.add(0xe4) // RETF
+  terminatingOpcodes.add(0xe5) // JUMPF
+
+  terminatingOpcodes.add(0xe0) // RJUMPing back into code section is OK
+
+  for (const opcode of terminatingOpcodes) {
+    if (!opcodeNumbers.has(opcode)) {
+      terminatingOpcodes.delete(opcode)
+    }
+  }
+
+  const validJumps = new Set<number>()
+
+  // Add all reachable code sections
+  const reachableSections: { [key: number]: Set<number> } = {}
+
+  let codeSection = -1
+  for (const code of container.body.codeSections) {
+    codeSection++
+
+    reachableSections[codeSection] = new Set()
+
+    const returningFunction = container.body.typeSections[codeSection].outputs === 0x80
```

Comment:
The naming of this variable is opposite of what is actually contained.  Only non-returning functions have an outputs section of 0x80

```suggestion
    // 0x80 indicates that this code section is a function that does not return a value
    const nonReturningFunction = container.body.typeSections[codeSection].outputs === 0x80
```